src/Tools/ExternalAccess/LiveShare/Tagger/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -0,0 +1,256 @@
+                }
+            }
+
+            internal void IncrementReferenceCount()
+            {
+                this._taggerReferenceCount++;
+            }
+
+            internal void DecrementReferenceCountAndDisposeIfNecessary()
+            {
+                this._taggerReferenceCount--;
+
+                if (this._taggerReferenceCount == 0)
+                {
+                    DisconnectFromWorkspace();
+                    this._workspaceRegistration.WorkspaceChanged -= OnWorkspaceRegistrationChanged;
+                    this._taggerProvider.DisconnectTagComputer(this._textBuffer);
+                }
+            }
+
+            private void ConnectToWorkspace(Workspace workspace)
+            {
+                workspace.WorkspaceChanged += this.OnWorkspaceChanged;
+                workspace.DocumentOpened += this.OnDocumentOpened;
+                workspace.DocumentActiveContextChanged += this.OnDocumentActiveContextChanged;
+                this._workspace = workspace;
+
+                EnqueueProcessSnapshotAsync();
+            }
+
+            public void DisconnectFromWorkspace()
+            {
+                var workspace = this._workspace;
+                if (workspace != null)
+                {
+                    workspace.WorkspaceChanged -= this.OnWorkspaceChanged;
+                    workspace.DocumentOpened -= this.OnDocumentOpened;
+                    workspace.DocumentActiveContextChanged -= this.OnDocumentActiveContextChanged;
+
+                    workspace = null;
+                }
+            }
+
+            private void EnqueueProcessSnapshotAsync(DocumentId updatedDocumentId = null)
+            {
+                var workspace = this._workspace;
+                if (workspace == null)
+                {
+                    return;
+                }
+
+                var documentId = workspace.GetDocumentIdInCurrentContext(this._textBuffer.AsTextContainer());
+                if (documentId == null)
+                {
+                    return;
+                }
+
+                // If the caller specified a DocumentId and it's the one for this buffer (in the current context),
+                // then there's nothing to do.
+                if (updatedDocumentId != null && updatedDocumentId != documentId)
+                {
+                    // This is very common and not very interesting, so don't record it.
+                    return;
+                }
+
+                if (Interlocked.CompareExchange(ref this._isRequestPending, 1, 0) != 0)
+                {
+                    return;
+                }
+
+                _workQueue.EnqueueBackgroundTask(c => this.EnqueueProcessSnapshotWorkerAsync(documentId, c), GetType() + "." + nameof(EnqueueProcessSnapshotAsync) + ".1", CancellationToken.None);
+            }
+
+            private async Task EnqueueProcessSnapshotWorkerAsync(DocumentId documentId, CancellationToken cancellationToken)
```

Comment:
>EnqueueProcessSnapshotWorkerAsync [](start = 31, length = 33)

I think the Roslyn code I borrowed this from had some additional side-effects for updating the workspace and/or documents.  They weren't interesting to TS, but you might want to confirm that they aren't important for Roslyn.